codelists/models.py

```diff
@@ -99,24 +99,36 @@ def coding_system(self):
 
     @cached_property
     def table(self):
-        return list(csv.reader(StringIO(self.csv_data)))
+        if self.csv_data:
+            return list(csv.reader(StringIO(self.csv_data)))
+        else:
+            code_to_term = self.coding_system.code_to_term(self.codes)
+            rows = [["code", "term"]]
+            rows.extend(
+                [code, code_to_term.get(code, "[Unknown]")] for code in self.codes
+            )
+            return rows
 
     @cached_property
     def codes(self):
-        if self.coding_system_id in ["ctv3", "ctv3tpp", "snomedct"]:
-            headers, *rows = self.table
-
-            for header in ["CTV3ID", "CTV3Code", "ctv3_id", "snomedct_id", "id"]:
-                if header in headers:
-                    ix = headers.index(header)
-                    break
-            else:
-                if self.codelist.slug == "ethnicity":
-                    ix = 1
+        if self.csv_data:
+            if self.coding_system_id in ["ctv3", "ctv3tpp", "snomedct"]:
+                headers, *rows = self.table
+
+                for header in ["CTV3ID", "CTV3Code", "ctv3_id", "snomedct_id", "id"]:
+                    if header in headers:
+                        ix = headers.index(header)
+                        break
                 else:
-                    ix = 0
+                    if self.codelist.slug == "ethnicity":
+                        ix = 1
+                    else:
+                        ix = 0
+
+                return tuple(sorted({row[ix] for row in rows}))
 
-            return tuple(sorted({row[ix] for row in rows}))
+        else:
+            return tuple(sorted(self.code_objs.values_list("code", flat=True)))
```

Comment:
You could flip this to short-circuit on the negative of the conditional, which I think would be clearer, and save some indentation on the positive part of the condition.

Reply:
Thanks.  Have addressed in 0bc934f1a2b3a86e8c592f0b544792172c442c08.